### PR TITLE
Add Fluent API support for FINDER and BATCH_FINDER methods.

### PR DESCRIPTION
**Pull request by karthikbalasub**
- Adds support for FINDER and BATCH_FINDER methods.
- Supports returning metadata and paging data with projection support.
- Includes support for finders with association key param (for association resources).

### CODE REVIEW

File: restli-tools/src/main/resources/macros/library.vm [hunk 1]
```diff
@@ -109,21 +109,34 @@
...
+      $assocKey.declaredType $assocKey.name#if( $foreach.hasNext || $hasMoreParams),#end
+    #end
+#end
+
 #macro(methodParams $method $includeOptional)
-  #set($hasOptionalParams = $includeOptional && ($method.optionalParams.size() > 0 || $method.supportedProjectionParams.size() > 0))
-  #foreach($param in $method.requiredParams)
+  #set($hasOptionalParams = $includeOptional && ($method.hasOptionalParams() || $method.hasProjectionParams()))
+  #foreach($param in $method.requiredParameters)
     $param.paramClassDisplayName $param.paramName#if( $foreach.hasNext || $hasOptionalParams),#end
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply restli-tools/src/main/resources/macros/library.vm:122]: I thought about it yesterday, but forgot to leave a comment. Left now :)

File: restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java [hunk 2]
```diff
@@ -120,33 +118,32 @@ public static String nameCamelCase(String name)
 
   public static String restMethodToClassPrefix(String name)
   {
-    ResourceMethod method = ResourceMethod.fromString(name);
-    switch (method)
+    switch (name)
     {
-      case GET:
+      case "GET":
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java:123]: This is called from the templates and can be used for finders/batch finders, where the method name is actual finder name.

[the remark this answers is not in the document]
Comment (junchuanwang) [reply restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java:123]: Looks like this method. the passed in argument is actually lower case.

[the remark this answers is not in the document]
Comment (karthikbalasub) [reply restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java]: You are correct, fixed it now.

[the remark this answers is not in the document]
Comment (junchuanwang) [reply restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java]: I am working on the rebase. I will fix this

File: restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java [hunk 3]
```diff
@@ -120,33 +118,24 @@ public static String nameCamelCase(String name)
 
   public static String restMethodToClassPrefix(String name)
   {
-    ResourceMethod method = ResourceMethod.fromString(name);
-    switch (method)
+    switch (name)
```
Comment (junchuanwang) [inline restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java:121]: I thought you can just use name.toUpperCase()

File: restli-tools/src/main/resources/apiVmTemplates/action.vm [hunk 4]
```diff
@@ -23,24 +23,14 @@ Copyright (c) 2021 LinkedIn Corp.
     @SuppressWarnings("unchecked")
     public CompletionStage<${method.valueClassDisplayName}> ${util.nameCamelCase(${method.name})}(
       #if(${method.isEntityAction()})
-      $spec.keyClassName $spec.idName,
+        $spec.keyClassName $spec.idName,
```
Comment (junchuanwang) [inline restli-tools/src/main/resources/apiVmTemplates/action.vm:26]: this line should also be guarded by "$method.hasActionParams()"

File: restli-tools/src/main/resources/apiVmTemplates/action.vm [hunk 5]
```diff
@@ -73,22 +63,12 @@ Copyright (c) 2021 LinkedIn Corp.
       #if(${method.isEntityAction()})
       $spec.keyClassName $spec.idName,
```
Comment (junchuanwang) [inline restli-tools/src/main/resources/apiVmTemplates/action.vm:64]: same here, for action without any params

**Review by junchuanwang — COMMENTED** [review]
ship with comments.

**Review by junchuanwang — APPROVED** [review]

File: restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java (hunk 3 above)
Comment (karthikbalasub) [reply restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/fluentspec/SpecUtils.java:121]: It doesn't matter as we are comparing the string directly.

File: restli-tools/src/main/resources/apiVmTemplates/action.vm (hunk 4 above)
Comment (karthikbalasub) [reply restli-tools/src/main/resources/apiVmTemplates/action.vm:26]: The hasRequiredParams check in line 22 ensures there will be params.

File: restli-tools/src/main/resources/apiVmTemplates/action.vm (hunk 5 above)
Comment (karthikbalasub) [reply restli-tools/src/main/resources/apiVmTemplates/action.vm:64]: same here, check on line 60 gurads this.

File: restli-tools/src/main/resources/macros/library.vm [hunk 6]
```diff
@@ -169,21 +177,34 @@
   #end
 #end
 
+#macro(assocKeyParams $method $includeOptional)
+    #set($hasMoreParams = $method.hasRequiredParams() || ($includeOptional && ($method.hasOptionalParams() || $method.hasProjectionParams())))
+    #foreach($assocKey in $method.assocKeys)
+      $assocKey.declaredType $assocKey.name#if( $foreach.hasNext || $hasMoreParams),#end
```
Comment (junchuanwang) [inline restli-tools/src/main/resources/macros/library.vm:183]: I actually think it should be "assocKey.bindingType" here. Can you please check?

**Pull request merged**